ompi/datatype/ompi_datatype_internal.h

```diff
@@ -570,16 +571,8 @@ extern const ompi_datatype_t* ompi_datatype_basicDatatypes[OMPI_DATATYPE_MPI_MAX
 #define OMPI_DATATYPE_INITIALIZER_UNSIGNED            OPAL_DATATYPE_INITIALIZER_UINT8
 #endif
 
-#if SIZEOF_LONG == 4
-#define OMPI_DATATYPE_INITIALIZER_LONG                OPAL_DATATYPE_INITIALIZER_INT4
-#define OMPI_DATATYPE_INITIALIZER_UNSIGNED_LONG       OPAL_DATATYPE_INITIALIZER_UINT4
-#elif SIZEOF_LONG == 8
-#define OMPI_DATATYPE_INITIALIZER_LONG                OPAL_DATATYPE_INITIALIZER_INT8
-#define OMPI_DATATYPE_INITIALIZER_UNSIGNED_LONG       OPAL_DATATYPE_INITIALIZER_UINT8
-#elif SIZEOF_LONG == 16
-#define OMPI_DATATYPE_INITIALIZER_LONG                OPAL_DATATYPE_INITIALIZER_INT16
-#define OMPI_DATATYPE_INITIALIZER_UNSIGNED_LONG       OPAL_DATATYPE_INITIALIZER_UINT16
-#endif
+#define OMPI_DATATYPE_INITIALIZER_LONG                OPAL_DATATYPE_INITIALIZER_LONG
```

Comment:
The idea was not to pollute OPAL with types that don't have a well defined size, and delegate all that mess up in the MPI layer.

Reply:
I agree except I don't see how else to handle it.  I don't like adding more OPAL_DATATYPE_* entries, but I didn't see any other mechanism that would allow pack/unpack to figure out that an MPI_LONG had been passed in and it therefore needed to pack it as 4 bytes (for external32).  The type construction of MPI_LONG maps to OPAL_DATATYPE_INT8 and after that the opal pack/unpack routines have no way to know whether an incoming OPAL_DATATYPE_INT8 came from an MPI_LONG or an MPI_INT8_T.

I think that's the reason OPAL_DATATYPE_BOOL exists when it could just be OPAL_DATATYPE_INT1.

I'm considering the possibility that ```opal_datatype_add(,pdtAdd,,,)``` might be able to see pdtAdd == ompi_mpi_long and where it does
```
        pLast->elem.common.type      = pdtAdd->id;
```
add another field like
```
        pLast->elem.common.ext32sz = 4;
```
but I'm not confident whether that would catch everything.  I don't think opal_datatype_add() is exclusively given base types.  Is that a plausible avenue to pursue?